### PR TITLE
CASSANDRA-20851 nodetool history

### PR DESCRIPTION
**Pull request by smiklosovic**
Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

### CODE REVIEW

File: test/unit/org/apache/cassandra/tools/nodetool/HistoryTest.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import org.junit.Test;
+
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.tools.Output;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class HistoryTest
```
Comment (smiklosovic) [inline test/unit/org/apache/cassandra/tools/nodetool/HistoryTest.java:33]: I have not extended anything and haven't mocked either. Currently, path to history file is wired and I wanted to have a way how to override that for test purposes. Once done like that, I realized I do not need anything else and I just test it by instantiating `History` and that's it.

I did not want to introduce system property either because if you think about what this does, it will print the content of a file. So an attacker might just set a system property by which it would be possible to print whatever file they want ...

Comment (Mmuzaf) [reply test/unit/org/apache/cassandra/tools/nodetool/HistoryTest.java:33]: I think we can execute the command `sjk` with some options. This should be quite safe and fast, and it shouldn't require any Cassandra nodes. Then, execute the history command to check that the SDK command is recorded in the history file. As a result we probably don't need a temp history file.

Did I get the problem correctly?

Comment (smiklosovic) [reply test/unit/org/apache/cassandra/tools/nodetool/HistoryTest.java:33]: @Mmuzaf 

No, I do not think so. You want me to execute "nodetool sjk" and then look into ~/.cassandra/nodetool.history file? I do not think this is a good idea because as I said, the path to history file is wired right now. So as you test, a lot of tests can execute nodetool commands. So all these commands would be visible in that history file, right?

I do not think this is isolated enough. Test suite is basically changing the content of this file as tests are progressing. So it is changing "under the hands".

For that reason I want to have a way how to completely isolate the history file and have its content under control to make further assertions on that (various edge conditions etc as shown in methods below).

Comment (Mmuzaf) [reply test/unit/org/apache/cassandra/tools/nodetool/HistoryTest.java:33]: What if we do the following: 
we can use `history` command with `--number-of-commands=1` and assert on the history command output (check the logs), can't we?  As far as I understand,  the tests are run one by one within a split run on CI. Therefore, it should be quite safe to validate only the last executed command. 
Don't need to access directly the history file.

Comment (smiklosovic) [reply test/unit/org/apache/cassandra/tools/nodetool/HistoryTest.java:33]: What about all the rest of the tests in this class which test various conditions. I am specifically testing like what happens when the file is empty, or that file does not exist, or I want to display specific number of lines etc ... I do not quite get what we are trying to do by what you suggest. Why do you insist on that, what are the actual benefits? The content of the file is under control and we are the ones telling what its content will be like. Self-encapsulated test.

**Review by Mmuzaf — APPROVED** [review]
+1 TY
